File: kreuzberg/packages/elixir/lib/kreuzberg/code_chunk.ex
defmodule Kreuzberg.CodeChunk do
  @moduledoc """
  Code chunk with source span and optional parent context.

  ## Fields

    * `:content` - Chunk text content
    * `:language` - Programming language
    * `:span` - Source span
    * `:context` - Optional parent context
  """

  @type t :: %__MODULE__{
          content: String.t(),
          language: String.t(),
          span: Kreuzberg.CodeSpan.t(),
          context: Kreuzberg.CodeChunkContext.t() | nil
        }

  defstruct content: "",
            language: "",
            span: %Kreuzberg.CodeSpan{},
            context: nil

  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    span =
      case Map.get(data, "span", Map.get(data, :span)) do
        %Kreuzberg.CodeSpan{} = s -> s
        map when is_map(map) -> Kreuzberg.CodeSpan.from_map(map)
        _ -> %Kreuzberg.CodeSpan{}
      end

    context =
      case Map.get(data, "context", Map.get(data, :context)) do
        %Kreuzberg.CodeChunkContext{} = c -> c
        map when is_map(map) -> Kreuzberg.CodeChunkContext.from_map(map)
        _ -> nil
      end

    %__MODULE__{
      content: Map.get(data, "content", Map.get(data, :content, "")),
      language: Map.get(data, "language", Map.get(data, :language, "")),
      span: span,
      context: context
    }
  end

  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = chunk) do
    %{
      "content" => chunk.content,
      "language" => chunk.language,
      "span" => Kreuzberg.CodeSpan.to_map(chunk.span),
      "context" =>
        if chunk.context do
          Kreuzberg.CodeChunkContext.to_map(chunk.context)
        end
    }
  end
end

File: kreuzberg/packages/php/src/Types/CodeChunk.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Code chunk with source span and optional parent context.
 *
 * @property-read string $content Chunk text content
 * @property-read string $language Programming language
 * @property-read CodeSpan $span Source span
 * @property-read CodeChunkContext|null $context Optional parent context
 */
readonly class CodeChunk
{
    public function __construct(
        public string $content,
        public string $language,
        public CodeSpan $span,
        public ?CodeChunkContext $context = null,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var string $content */
        $content = $data['content'] ?? '';

        /** @var string $language */
        $language = $data['language'] ?? '';

        /** @var array<string, mixed> $spanData */
        $spanData = $data['span'] ?? [];

        /** @var array<string, mixed>|null $contextData */
        $contextData = is_array($data['context'] ?? null) ? $data['context'] : null;

        return new self(
            content: $content,
            language: $language,
            span: CodeSpan::fromArray($spanData),
            context: $contextData !== null ? CodeChunkContext::fromArray($contextData) : null,
        );
    }
}
